torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
 
         else:
             raise NotImplementedError("Method not yet implemented")
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
```

Comment:
fyi: Linear has broadcasting behavior because it dispatches to matmul https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/Linear.cpp.html#37

Not sure if `Linear` the module is different but:

```
>>> r = torch.rand
>>> torch._C._nn.linear(r(4), r(8, 4), r(8)).size()
torch.Size([8])
```

Reply:
I will check this! I went based on the documentation here https://pytorch.org/docs/stable/generated/torch.nn.Linear.html which talked about the input and output shapes.

Reply:
As you've written it I think is fine for `nn.Linear` the Module..